include/swift/Sema/ConstraintSystem.h

```diff
@@ -908,15 +908,22 @@ using OpenedTypeMap =
 /// within a constraint system.
 struct ContextualTypeInfo {
   TypeLoc typeLoc;
+
+  /// The contextual type as it should be used in the actual constraint system.
+  /// This will have any placeholder types replaced with type variables, unbound
+  /// generics opened, etc.
+  Optional<Type> typeForConstraint;
```

Comment:
I'd suggest we move this to `ConstraintSystem::ContextualTypes` and make that map's "value" to be a pair of immutable `ContextualTypeInfo` and a `Type`. This way this type doesn't have to be propagated to a solution and we keep "opening" logic local to the constraint system which is the only place that needs to know about that. Doing so also makes it simpler to work with, because value could be updated by reference...